### PR TITLE
Implement "use constructor" code action

### PR DESCRIPTION
**Pull request by isovector**
This PR adds a `Use constructor <x>` code action. It only runs on data types with small numbers of constructors (not `HsExpr`) in order to not spam the code action list.

Fixes https://github.com/isovector/haskell-language-server/issues/21
Subsumes #521

### CODE REVIEW

**Review by pepeiborra — APPROVED** [review]

File: plugins/hls-tactics-plugin/src/Ide/Plugin/Tactic/LanguageServer/TacticProviders.hs [hunk 1]
```diff
@@ -71,6 +75,25 @@ commandProvider HomomorphismLambdaCase =
   requireExtension LambdaCase $
     filterGoalType ((== Just True) . lambdaCaseable) $
       provide HomomorphismLambdaCase ""
+commandProvider UseDataCon =
+  requireFeature FeatureUseDataCon $
+    filterTypeProjection
+        ( guardLength (<= 5)
```
Comment (pepeiborra) [inline plugins/hls-tactics-plugin/src/Ide/Plugin/Tactic/LanguageServer/TacticProviders.hs:81]: `5` should be configurable

Comment (isovector) [reply plugins/hls-tactics-plugin/src/Ide/Plugin/Tactic/LanguageServer/TacticProviders.hs:81]: Do you think anyone would actually ever change this value, though?

**Comment by googleson78** [comment]
Does this currently do any type checking on the suggested constructors? e.g. if the type indices that the constructor uses will match up with the required type. Is it possible to easily do?

**Comment by isovector** [comment]
@pepeiborra implemented configurable option. PTAL

**Comment by isovector** [comment]
@googleson78 oh, I didn't even think of that. Great idea. I'll implement it here.

File: plugins/hls-tactics-plugin/src/Ide/Plugin/Tactic/TestTypes.hs [hunk 2]
```diff
@@ -33,20 +34,23 @@ tacticTitle UseDataCon dcon          = "Use constructor " <> dcon
 
 ------------------------------------------------------------------------------
 -- | Plugin configuration for tactics
-newtype Config = Config
-  { cfg_feature_set :: FeatureSet
+data Config = Config
+  { cfg_feature_set          :: FeatureSet
+  , cfg_max_use_ctor_actions :: Int
```
Comment (pepeiborra) [inline plugins/hls-tactics-plugin/src/Ide/Plugin/Tactic/TestTypes.hs:39]: You might want to make the new field optional to allow for partial configs

Comment (isovector) [reply plugins/hls-tactics-plugin/src/Ide/Plugin/Tactic/TestTypes.hs:39]: Good catch. But I'd rather use a loose parser than juggle maybes everywhere. Will implement.

**Comment by isovector** [comment]
@googleson78 done. Mind checking the tests in 3d79859 and making sure you agree with them?

**Comment by googleson78** [comment]
They seem ok. Nice that other tactics machinery can be reused here!

**Pull request merged**